### ISSUE TITLE
[Feature] Select singular tests by tag

### ISSUE THREAD

**Issue by remisalmon** [dbt-labs/dbt-core#8138]
### Is this your first time submitting a feature request?

- [X] I have read the [expectations for open source contributors](https://docs.getdbt.com/docs/contributing/oss-expectations)
- [X] I have searched the existing issues, and I could not find an existing issue for this feature
- [X] I am requesting a straightforward extension of existing dbt functionality, rather than a Big Idea better suited to a discussion

### Describe the feature

It is currently not possible to select singular tests by tag and `dbt test --select test_type:singular,tag:my_tag` will run singular tests whose models are tagged with `my_tag` as opposed to singular tests tagged with `my_tag`.

This is documented at https://docs.getdbt.com/reference/resource-configs/tags#definition but makes tags on singular tests confusing and not very useful.

### Describe alternatives you've considered

The only workaround is to explicitly select singular tests by name which does not scale well (at all).

### Who will this benefit?

Anyone using singular tests who wants to run a subset of those by filtering by tag.

### Are you interested in contributing this feature?

No.

### Anything else?

See also this same issue here: https://stackoverflow.com/questions/75931209/how-to-add-tag-to-singular-test-in-dbt

**Comment by dbeatty10** [comment]
Thanks for raising this @remisalmon !

@graciegoheen, @jtcohen6, and I briefly discussed this a few minutes ago.

### TLDR
We agree with you that this would be useful!

### More detail
Indeed, the docs say that the following will indirectly run all tests associated with the models that are tagged:

```shell
dbt test --select tag:my_tag
```

But we agree that it would be great if it could say this instead:

`dbt test --select tag:my_tag` (indirectly runs all tests associated with the models that are tagged as well as any singular tests that are tagged)

### Next steps
I don't know how soon we'd be able to get to this ourselves, so I'm going to mark this as `help_wanted` in the meantime.

**Comment by Paullee6997** [comment]
Hi @remisalmon @dbeatty10,

Jumping in here due to curiosity, but was hoping to get some clarification on exactly the unexpected behavior is and what we would like as an alternative.

Ran through some examples locally after forking off of dbt-core v1.6.2 and am seeing the following behavior. 

### Example Scenario
- We have two singular tests named `singular_one, singular_two`, where each is tagged with `tag_one`
- We have a model `my_first_dbt_model` with the same tag `tag_one`
- We have a pre-built test `not_null` applied on one of the columns in our model `my_first_dbt_model`

When running `dbt test --select tag:tag_one`
- We have all the tests built being run

When running `dbt test --select test_type:singular, tag:tag_one`
- We have all only the singulars tests `singular_one` and `singular_two` being run

The behavior listed above, is to my expectations appropriate such that by specifying `test_type:singular | generic` dbt will only select the tests that match the specified test_type regardless of whether models share the same tag as tests. Would really appreciate any clarification on anything I might be missing or not catching with the initial request.

**Comment by aahel** [comment]
Closing this as resolved — selecting singular tests by tag works correctly in the current version of dbt-core.
